# 대기환경 데이터 전처리
- numpy와 pandas를 활용해 충청북도의 대기환경 데이터를 가공
- 데이터 출처: 에어코리아

In [122]:
import pandas as pd
import numpy as np

'air_19_1.csv' 파일을 데이터프레임으로 읽어주세요. encoding은 'CP949'입니다.

In [123]:
air = pd.read_csv('air_19_1.csv', encoding = 'CP949')
air.head()

지역     망   측정소코드 측정소명        측정일시     SO2   CO     O3     NO2  PM10  \
0  충북 괴산군  교외대기  633361  청천면  2019012906  0.0001  0.4  0.017  0.0048  49.0   
1  충북 괴산군  교외대기  633361  청천면  2019012909  0.0001  0.4  0.017  0.0051  46.0   
2  충북 괴산군  교외대기  633361  청천면  2019012904  0.0002  0.5  0.018  0.0051  63.0   
3  충북 괴산군  교외대기  633361  청천면  2019012908  0.0002  0.4  0.016  0.0058  47.0   
4  충북 괴산군  교외대기  633361  청천면  2019012905  0.0002  0.5  0.017  0.0053  58.0   

   PM25                  주소  
0   NaN  충북 괴산군 청천면 송문로 덕평길  
1   NaN  충북 괴산군 청천면 송문로 덕평길  
2   NaN  충북 괴산군 청천면 송문로 덕평길  
3   NaN  충북 괴산군 청천면 송문로 덕평길  
4   NaN  충북 괴산군 청천면 송문로 덕평길

결측값을 확인해주세요

In [124]:
air.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15134 entries, 0 to 15133
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   지역      15134 non-null  object 
 1   망       15134 non-null  object 
 2   측정소코드   15134 non-null  int64  
 3   측정소명    15134 non-null  object 
 4   측정일시    15134 non-null  int64  
 5   SO2     11735 non-null  float64
 6   CO      11707 non-null  float64
 7   O3      12494 non-null  float64
 8   NO2     12391 non-null  float64
 9   PM10    12347 non-null  float64
 10  PM25    11766 non-null  float64
 11  주소      15134 non-null  object 
dtypes: float64(6), int64(2), object(4)
memory usage: 1.4+ MB


모든 결측값을 0으로 바꿔주세요.

In [125]:
air = air.fillna(0)
air.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15134 entries, 0 to 15133
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   지역      15134 non-null  object 
 1   망       15134 non-null  object 
 2   측정소코드   15134 non-null  int64  
 3   측정소명    15134 non-null  object 
 4   측정일시    15134 non-null  int64  
 5   SO2     15134 non-null  float64
 6   CO      15134 non-null  float64
 7   O3      15134 non-null  float64
 8   NO2     15134 non-null  float64
 9   PM10    15134 non-null  float64
 10  PM25    15134 non-null  float64
 11  주소      15134 non-null  object 
dtypes: float64(6), int64(2), object(4)
memory usage: 1.4+ MB


SO2, CO, O3, NO2, PM10, PM25 컬럼의 이름을 아황산가스, 일산화탄소, 오존, 이산화질소, 미세먼지, 초미세먼지로 바꿔주세요

In [126]:
air = air.rename({'SO2' : '아황산가스', 'CO' : '일산화탄소', 'O3' : '오존', 'NO2' : '이산화질소', 'PM10' : '미세먼지', 'PM25' : '초미세먼지'}, axis = 1)
air.columns

Index(['지역', '망', '측정소코드', '측정소명', '측정일시', '아황산가스', '일산화탄소', '오존', '이산화질소',
       '미세먼지', '초미세먼지', '주소'],
      dtype='object')

측정일시 컬럼에서 연도, 월, 일을 추출해 연도, 월, 일 컬럼을 새로 생성해주세요. (힌트: map 사용)

In [127]:
air['연도'] = air.측정일시.astype('str').map(lambda x : x[: 4])
air['월'] = air.측정일시.astype('str').map(lambda x : x[4 : 6])
air['일'] = air.측정일시.astype('str').map(lambda x : x[6 : 8])
air.columns

Index(['지역', '망', '측정소코드', '측정소명', '측정일시', '아황산가스', '일산화탄소', '오존', '이산화질소',
       '미세먼지', '초미세먼지', '주소', '연도', '월', '일'],
      dtype='object')

air에서 '측정소명', '측정일시','주소' 컬럼을 삭제해주세요.

In [128]:
air = air.drop(['측정소명', '측정일시', '주소'], axis = 1)
air.columns

Index(['지역', '망', '측정소코드', '아황산가스', '일산화탄소', '오존', '이산화질소', '미세먼지', '초미세먼지',
       '연도', '월', '일'],
      dtype='object')

air 데이터를 측정소코드, 연도, 월, 일을 기준으로 오름차순 정렬해주세요.

In [129]:
air = air.sort_values(by = ['측정소코드', '연도', '월', '일'], axis = 0)
air.head(10)

지역     망   측정소코드  아황산가스  일산화탄소     오존  이산화질소  미세먼지  초미세먼지    연도   월  \
4803  충북 청주시  도시대기  533112  0.004    0.3  0.024  0.012  35.0   28.0  2019  01   
8009  충북 청주시  도시대기  533112  0.005    0.5  0.023  0.012  41.0   35.0  2019  01   
8015  충북 청주시  도시대기  533112  0.005    0.4  0.020  0.013  42.0   34.0  2019  01   
8016  충북 청주시  도시대기  533112  0.005    0.5  0.021  0.012  48.0   31.0  2019  01   
8035  충북 청주시  도시대기  533112  0.005    0.4  0.018  0.013  34.0   25.0  2019  01   
8058  충북 청주시  도시대기  533112  0.005    0.8  0.003  0.019  59.0   40.0  2019  01   
8066  충북 청주시  도시대기  533112  0.005    0.8  0.003  0.018  54.0   36.0  2019  01   
8121  충북 청주시  도시대기  533112  0.005    0.4  0.029  0.010  24.0   14.0  2019  01   
8179  충북 청주시  도시대기  533112  0.005    0.3  0.027  0.012  19.0   17.0  2019  01   
8188  충북 청주시  도시대기  533112  0.005    0.4  0.026  0.011  21.0   12.0  2019  01   

       일  
4803  01  
8009  01  
8015  01  
8016  01  
8035  01  
8058  01  
8066  01  
8121  01  
8179  01  
8188  01

정렬을 하면서 인덱스가 뒤섞였습니다. 데이터의 인덱스를 0부터 15133까지 숫자로 초기화해주세요.

In [130]:
air = air.reset_index()
air.head()

index      지역     망   측정소코드  아황산가스  일산화탄소     오존  이산화질소  미세먼지  초미세먼지    연도  \
0   4803  충북 청주시  도시대기  533112  0.004    0.3  0.024  0.012  35.0   28.0  2019   
1   8009  충북 청주시  도시대기  533112  0.005    0.5  0.023  0.012  41.0   35.0  2019   
2   8015  충북 청주시  도시대기  533112  0.005    0.4  0.020  0.013  42.0   34.0  2019   
3   8016  충북 청주시  도시대기  533112  0.005    0.5  0.021  0.012  48.0   31.0  2019   
4   8035  충북 청주시  도시대기  533112  0.005    0.4  0.018  0.013  34.0   25.0  2019   

    월   일  
0  01  01  
1  01  01  
2  01  01  
3  01  01  
4  01  01

groupby를 이용해 지역별 미세먼지 평균을 구하고, 구한 데이터프레임을 dust 변수에 할당하세요. 

In [131]:
dust = air.loc[:, ['지역', '미세먼지']].groupby('지역').mean()
dust

미세먼지
지역               
충북 괴산군  29.293011
충북 단양군  63.723118
충북 보은군  70.123656
충북 영동군  63.575269
충북 옥천군  63.645161
충북 음성군  54.445892
충북 제천시  69.536290
충북 증평군  68.663978
충북 진천군  68.592742
충북 청주시  55.501680
충북 충주시  69.220430

dust의 인덱스는 '지역'입니다. 인덱스를 컬럼으로 빼주세요.

In [132]:
dust = dust.reset_index()
dust

지역       미세먼지
0   충북 괴산군  29.293011
1   충북 단양군  63.723118
2   충북 보은군  70.123656
3   충북 영동군  63.575269
4   충북 옥천군  63.645161
5   충북 음성군  54.445892
6   충북 제천시  69.536290
7   충북 증평군  68.663978
8   충북 진천군  68.592742
9   충북 청주시  55.501680
10  충북 충주시  69.220430

dust에서, 충북 전체의 미세먼지 평균 초과의 지역만 추출해 dust2에 할당해주세요.

In [133]:
dust2 = dust[dust['미세먼지'] > dust['미세먼지'].mean()]
dust2

지역       미세먼지
1   충북 단양군  63.723118
2   충북 보은군  70.123656
3   충북 영동군  63.575269
4   충북 옥천군  63.645161
6   충북 제천시  69.536290
7   충북 증평군  68.663978
8   충북 진천군  68.592742
10  충북 충주시  69.220430

groupby를 이용해 지역별 오존 평균을 구하고, 구한 데이터프레임을 ozon 변수에 할당하세요.

In [134]:
ozon = air.loc[:, ['지역', '오존']].groupby('지역').mean()
ozon

오존
지역              
충북 괴산군  0.012702
충북 단양군  0.013203
충북 보은군  0.019886
충북 영동군  0.013949
충북 옥천군  0.013048
충북 음성군  0.010070
충북 제천시  0.013336
충북 증평군  0.014181
충북 진천군  0.017667
충북 청주시  0.009882
충북 충주시  0.014757

dust2와 ozon을 병합하되, 두 데이터에 모두 있는 지역만 병합해주세요. 

In [135]:
dust_ozon_mean = pd.merge(dust2, ozon, left_on = '지역', right_index = True, how = 'inner')
dust_ozon_mean

지역       미세먼지        오존
1   충북 단양군  63.723118  0.013203
2   충북 보은군  70.123656  0.019886
3   충북 영동군  63.575269  0.013949
4   충북 옥천군  63.645161  0.013048
6   충북 제천시  69.536290  0.013336
7   충북 증평군  68.663978  0.014181
8   충북 진천군  68.592742  0.017667
10  충북 충주시  69.220430  0.014757

병합한 데이터를 한글이 깨지지 않게 csv로 저장해주세요. 

In [136]:
dust_ozon_mean.to_csv('dust_ozon_mean.csv', encoding = 'CP949')

인덱스는 일(day), 컬럼은 지역이고 값은 미세먼지의 평균인 피벗 테이블을 생성해주세요.

In [137]:
piv = pd.pivot_table(air, index = '일', columns = '지역', values = '미세먼지', aggfunc = 'mean')
piv.head()

지역     충북 괴산군     충북 단양군     충북 보은군     충북 영동군     충북 옥천군     충북 음성군  \
일                                                                      
01  19.770833  52.250000  42.500000  43.750000  42.458333  47.666667   
02  11.250000  41.500000  38.250000  39.916667  34.708333  38.208333   
03  14.937500  52.791667  47.125000  39.541667  42.083333  52.791667   
04  25.791667  62.875000  63.541667  47.250000  55.583333  63.458333   
05  31.437500  71.000000  76.666667  60.750000  62.083333  85.791667   

지역     충북 제천시     충북 증평군     충북 진천군     충북 청주시     충북 충주시  
일                                                          
01  54.291667  42.291667  36.916667  36.109375  53.854167  
02  52.708333  32.583333  33.375000  28.625000  46.979167  
03  57.041667  49.708333  47.541667  40.057292  53.312500  
04  71.875000  70.916667  64.625000  54.906250  89.437500  
05  82.750000  63.500000  69.041667  50.098958  71.458333

지역 컬럼을 기준으로 오름차순(가나다 순)으로 정렬해주세요.
(힌트: 정렬 기준은 piv.columns.tolist().sort() 사용 )

In [138]:
piv.columns.tolist().sort()
piv.head()

지역     충북 괴산군     충북 단양군     충북 보은군     충북 영동군     충북 옥천군     충북 음성군  \
일                                                                      
01  19.770833  52.250000  42.500000  43.750000  42.458333  47.666667   
02  11.250000  41.500000  38.250000  39.916667  34.708333  38.208333   
03  14.937500  52.791667  47.125000  39.541667  42.083333  52.791667   
04  25.791667  62.875000  63.541667  47.250000  55.583333  63.458333   
05  31.437500  71.000000  76.666667  60.750000  62.083333  85.791667   

지역     충북 제천시     충북 증평군     충북 진천군     충북 청주시     충북 충주시  
일                                                          
01  54.291667  42.291667  36.916667  36.109375  53.854167  
02  52.708333  32.583333  33.375000  28.625000  46.979167  
03  57.041667  49.708333  47.541667  40.057292  53.312500  
04  71.875000  70.916667  64.625000  54.906250  89.437500  
05  82.750000  63.500000  69.041667  50.098958  71.458333

In [139]:
piv = piv.sort_index(axis = 0)
piv.head()

지역     충북 괴산군     충북 단양군     충북 보은군     충북 영동군     충북 옥천군     충북 음성군  \
일                                                                      
01  19.770833  52.250000  42.500000  43.750000  42.458333  47.666667   
02  11.250000  41.500000  38.250000  39.916667  34.708333  38.208333   
03  14.937500  52.791667  47.125000  39.541667  42.083333  52.791667   
04  25.791667  62.875000  63.541667  47.250000  55.583333  63.458333   
05  31.437500  71.000000  76.666667  60.750000  62.083333  85.791667   

지역     충북 제천시     충북 증평군     충북 진천군     충북 청주시     충북 충주시  
일                                                          
01  54.291667  42.291667  36.916667  36.109375  53.854167  
02  52.708333  32.583333  33.375000  28.625000  46.979167  
03  57.041667  49.708333  47.541667  40.057292  53.312500  
04  71.875000  70.916667  64.625000  54.906250  89.437500  
05  82.750000  63.500000  69.041667  50.098958  71.458333

일별 미세먼지 평균을 피벳 테이블의 새로운 컬럼으로 추가해주세요.

In [140]:
piv['일별 미세먼지 평균'] = piv.mean(axis = 1)
piv.head()

지역     충북 괴산군     충북 단양군     충북 보은군     충북 영동군     충북 옥천군     충북 음성군  \
일                                                                      
01  19.770833  52.250000  42.500000  43.750000  42.458333  47.666667   
02  11.250000  41.500000  38.250000  39.916667  34.708333  38.208333   
03  14.937500  52.791667  47.125000  39.541667  42.083333  52.791667   
04  25.791667  62.875000  63.541667  47.250000  55.583333  63.458333   
05  31.437500  71.000000  76.666667  60.750000  62.083333  85.791667   

지역     충북 제천시     충북 증평군     충북 진천군     충북 청주시     충북 충주시  일별 미세먼지 평균  
일                                                                      
01  54.291667  42.291667  36.916667  36.109375  53.854167   42.896307  
02  52.708333  32.583333  33.375000  28.625000  46.979167   36.191288  
03  57.041667  49.708333  47.541667  40.057292  53.312500   45.175663  
04  71.875000  70.916667  64.625000  54.906250  89.437500   60.932765  
05  82.750000  63.500000  69.041667  50.098958  71.458333   65.870739

지역별 미세먼지 평균을 피벗 테이블의 마지막 행으로 추가해주세요

In [144]:
piv.loc['지역별 미세먼지 평균'] = piv.mean(axis = 0)
piv.tail()

지역              충북 괴산군     충북 단양군     충북 보은군     충북 영동군     충북 옥천군     충북 음성군  \
일                                                                               
28           39.729167  67.500000  74.041667  69.750000  68.791667  40.625000   
29           34.708333  72.250000  79.833333  80.791667  75.000000  41.333333   
30           32.437500  70.291667  76.250000  71.416667  65.625000  30.395833   
31           34.958333  77.625000  72.708333  66.916667  24.708333  39.458333   
지역별 미세먼지 평균  29.293011  63.723118  70.123656  63.575269  63.645161  62.514254   

지역              충북 제천시     충북 증평군     충북 진천군     충북 청주시     충북 충주시  일별 미세먼지 평균  
일                                                                               
28           65.125000  72.875000  78.625000  56.182292  73.416667   64.241951  
29           79.083333  74.041667  79.750000  62.635417  77.625000   68.822917  
30           72.333333  75.291667  65.375000  57.411458  79.437500   63.296875  
31           85.000000  66.875000  66.166667  51.718750  77.833333   60.360795  
지역별 미세먼지 평균  69.536290  68.663978  68.592742  55.501680  69.220430   62.217235

수고하셨습니다!